### PR TITLE
refactor: abstract hooker and sinker interface

### PR DESCRIPTION
**Pull request by tianzhou**
* Hooker registers the webhook handler at path `/foo`
* Sinker registers the webhook payload receiver at path `/foo`
* Configure external service foo to post webhook event to `/foo`

## Workflow

1. External service `foo` posts a webhook event to `/foo`
2. Hooker responds at `/foo`, extract the webhook payload, and performs some processing
3. Hooker sends the payload to Sinker registered at `/foo`

A specific case in this PR from GitHub webhook event to post messages to Lark group:

GitHub service webhook event -> GitHub Hooker -> Lark Sinker

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -78,49 +67,13 @@ func Execute() error {
 	return NewRootCmd().Execute()
 }
 
-func start() {
-	if flags.larkWebbookURLs == "" {
-		log.Fatal(`The "--lark-urls" or "LARK_URLS" is required`)
+func start(f *flamego.Flame) {
+	if err := hook.Prepare(); err != nil {
+		log.Fatalf("Failed to prepare hook: %v\n", err)
```
Comment (unknwon) [inline cmd/root.go:72]: FYI, `log.Fatalf` (or any `log.XYZf` from stdlib) always prints to a new line, below is effectively the same
```suggestion
		log.Fatalf("Failed to prepare hook: %v", err)
```

File: hook/github.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/bytebase/relay/payload"
+	"github.com/bytebase/relay/sink"
+	"github.com/flamego/flamego"
+	flag "github.com/spf13/pflag"
+)
+
+var (
+	refPrefix string
+	_         Hooker = (*hooker)(nil)
+)
```
Comment (unknwon) [inline hook/github.go:18]: nit: they are two types of vars (regular vs. compile check), putting them into a single group doesn't feel great.

File: hook/github.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+// NewGitHub creates a GitHub hooker
+func NewGitHub() Hooker {
+	return &hooker{}
+}
+
+type hooker struct {
+}
```
Comment (unknwon) [inline hook/github.go:26]: Shouldn't this be called `githubHooker`? I'd imagine we'll have `gitlabHooker` for example.

File: hook/hook.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/flamego/flamego"
+	"github.com/hashicorp/go-multierror"
+	flag "github.com/spf13/pflag"
+)
+
+// Hooker is the interface for the webhook originator
+type Hooker interface {
+	prepare() error
+	register(fs *flag.FlagSet, f *flamego.Flame, path string)
```
Comment (unknwon) [inline hook/hook.go:14]: Please add some docstrings.

File: hook/hook.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"sync"
+
+	"github.com/flamego/flamego"
+	"github.com/hashicorp/go-multierror"
+	flag "github.com/spf13/pflag"
+)
+
+// Hooker is the interface for the webhook originator
```
Comment (unknwon) [inline hook/hook.go:11]:
```suggestion
// Hooker is the interface for the webhook originator.
```

File: hook/hook.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+		panic("hook: Register hooker is nil")
+	}
+	if _, dup := hooks[path]; dup {
+		panic("hook: Register called twice for hooker " + path)
+	}
+	h.register(fs, f, path)
+	hooks[path] = h
+}
+
+// Prepare prepares all registered hooks before running (e.g. validate flags)
```
Comment (unknwon) [inline hook/hook.go:38]:
```suggestion
// Prepare prepares all registered hooks before running (e.g. validate flags).
```

File: payload/github.go [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package payload
+
+type GitHub struct {
```
Comment (unknwon) [inline payload/github.go:3]:
```suggestion
type GitHubPushEvent struct {
```

File: sink/lark.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	_           Sinker = (*sinker)(nil)
+)
+
+// NewLark creates a Lark sinker
+func NewLark() Sinker {
+	return &sinker{}
+}
+
+type sinker struct {
+}
```
Comment (unknwon) [inline sink/lark.go:30]: Ditto
```suggestion
type larkSinker struct {
}
```

File: sink/sink.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/flamego/flamego"
+	"github.com/hashicorp/go-multierror"
+	flag "github.com/spf13/pflag"
+)
+
+// Sinker is the interface for receiving the webhook payload from the Hooker
+type Sinker interface {
+	register(fs *flag.FlagSet)
+	prepare() error
+	send(c context.Context, path string, payload interface{}) error
```
Comment (unknwon) [inline sink/sink.go:16]: Please add docstrings

File: hook/hook.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+// Register registers the hook under path.
+// e.g. If you register the foo hook at /foo, then you go to service foo's webhook
+// setting page and configure the webhook to post events to <<Relay Host>>/foo.
+func Register(fs *flag.FlagSet, f *flamego.Flame, h Hooker, path string) {
+	hooksMu.Lock()
+	defer hooksMu.Unlock()
+	if h == nil {
+		panic("hook: Register hooker is nil")
+	}
```
Comment (unknwon) [inline hook/hook.go:30]: nit
```suggestion
	if h == nil {
		panic("hook: Register hooker is nil")
	}

	hooksMu.Lock()
	defer hooksMu.Unlock()
```

File: hook/hook.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+// Hooker is the interface for the webhook originator
+type Hooker interface {
+	prepare() error
+	register(fs *flag.FlagSet, f *flamego.Flame, path string)
+}
+
+var (
+	hooksMu sync.RWMutex
+	hooks   = make(map[string]Hooker)
```
Comment (unknwon) [inline hook/hook.go:19]: nit: My brain feels a bit split here
```suggestion
	hookersMu sync.RWMutex
	hookers   = make(map[string]Hooker)
```

WDYT?

File: sink/sink.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+var (
+	sinksMu sync.RWMutex
+	sinks   = make(map[string][]Sinker)
```
Comment (unknwon) [inline sink/sink.go:21]: Ditto

File: sink/sink.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+// Register registers the sinker for path.
+// e.g. If service foo posts the webhook event to path /foo,
+// You can register a sinker at path /foo to receive that webhook event payload.
+func Register(fs *flag.FlagSet, f *flamego.Flame, s Sinker, path string) {
+	sinksMu.Lock()
+	defer sinksMu.Unlock()
+	if s == nil {
+		panic("sink: Register sinker is nil")
+	}
+
```
Comment (unknwon) [inline sink/sink.go:33]: nit

```suggestion
	if s == nil {
		panic("sink: Register sinker is nil")
	}

	sinksMu.Lock()
	defer sinksMu.Unlock()
```

File: sink/sink.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+				panic("sink: Register called twice for sinker " + path)
+			}
+		}
+		s.register(fs)
+		sinks[path] = append(list, s)
+	} else {
+		s.register(fs)
+		sinks[path] = []Sinker{s}
+	}
+}
```
Comment (unknwon) [inline sink/sink.go:47]: Pretty sure you can safely iterate over and append to a nil slice.

```suggestion
	for _, item := range sinks[path] {
		if item == s {
			panic("sink: Register called twice for sinker " + path)
		}
	}
	s.register(fs)
	sinks[path] = append(sinks[path], s)
}
```

File: sink/sink.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+		for _, sink := range sinkList {
+			if err := sink.prepare(); err != nil {
+				result = multierror.Append(result, err)
+			}
+		}
+	}
+	return result
+}
+
+func Send(c context.Context, path string, payload interface{}) error {
```
Comment (unknwon) [inline sink/sink.go:65]: Please add docstring

File: sink/sink.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+		}
+		s.register(fs)
+		sinks[path] = append(list, s)
+	} else {
+		s.register(fs)
+		sinks[path] = []Sinker{s}
+	}
+}
+
+// Prepare prepares all registered sinks before running (e.g. validate flags)
```
Comment (unknwon) [inline sink/sink.go:49]:
```suggestion
// Prepare prepares all registered sinks before running (e.g. validate flags).
```

**Review by unknwon — CHANGES_REQUESTED** [review]
Overall LGTM, request changes because I want to make sure we sort out some details.

File: hook/github.go (hunk 3 above)
Comment (tianzhou) [reply hook/github.go:29]: this is already inside github.go

File: sink/lark.go (hunk 8 above)
Comment (tianzhou) [reply sink/lark.go:33]: this is already inside lark.go

File: hook/github.go (hunk 2 above)
Comment (tianzhou) [reply hook/github.go]: Done

File: sink/sink.go (hunk 12 above)
Comment (tianzhou) [reply sink/sink.go:26]: Done

File: hook/hook.go (hunk 4 above)
Comment (tianzhou) [reply hook/hook.go]: Done

**Review by tianzhou — COMMENTED** [review]
PTAL

File: cmd/root.go (hunk 1 above)
Comment (tianzhou) [reply cmd/root.go]: Done

File: hook/github.go (hunk 3 above)
Comment (tianzhou) [reply hook/github.go]: Fixed

**Review by unknwon — APPROVED** [review]

**Pull request merged**